src/main/java/org/jenkins/tools/test/PluginCompatTester.java

```diff
@@ -542,90 +402,74 @@ private TestExecutionResult testPluginAgainst(
+                    mconfig,
+                    pluginCheckoutDir,
+                    buildLogFile,
+                    "clean",
+                    "process-test-classes",
+                    "-Dmaven.javadoc.skip");
         }
+        ranCompile = true;
+
+        List<String> args = new ArrayList<>();
+        Map<String, String> userProperties = mconfig.userProperties;
+        args.add(
+                String.format(
+                        "--define=forkCount=%s", userProperties.getOrDefault("forkCount", "1")));
+        args.add("hpi:resolve-test-dependencies");
+        args.add("hpi:test-hpl");
+        args.add("surefire:test");
+
+        // Run preexecution hooks
+        List<String> testTypes = new LinkedList<>();
+        testTypes.add("surefire"); // default
+        Map<String, Object> forExecutionHooks = new HashMap<>();
+        forExecutionHooks.put("pluginName", plugin.name);
+        forExecutionHooks.put("plugin", plugin);
+        forExecutionHooks.put("args", args);
+        forExecutionHooks.put("pomData", pomData);
+        forExecutionHooks.put("pom", new MavenPom(pluginCheckoutDir));
+        forExecutionHooks.put("coreCoordinates", coreCoordinates);
+        forExecutionHooks.put("config", config);
+        forExecutionHooks.put("pluginDir", pluginCheckoutDir);
+        forExecutionHooks.put("types", testTypes);
+        pcth.runBeforeExecution(forExecutionHooks);
+        args = (List<String>) forExecutionHooks.get("args");
+        Set<String> types = new HashSet<>((List<String>) forExecutionHooks.get("types"));
+        userProperties.put("types", String.join(",", types));
+
+        // Execute with tests
+        runner.run(mconfig, pluginCheckoutDir, buildLogFile, args.toArray(new String[0]));
     }
 
-    public void cloneFromSCM(
-            PomData pomData, String name, String version, File checkoutDirectory, String tag) {
-        String scmTag = StringUtils.isNotEmpty(tag) ? tag : getScmTag(pomData, name, version);
-        String connectionURLPomData = pomData.getConnectionUrl();
+    public static void cloneFromScm(
+            String url, String fallbackGitHubOrganization, String scmTag, File checkoutDirectory)
+            throws PluginSourcesUnavailableException {
         List<String> connectionURLs = new ArrayList<>();
-        connectionURLs.add(connectionURLPomData);
-        if (config.getFallbackGitHubOrganization() != null) {
+        connectionURLs.add(url);
+        if (fallbackGitHubOrganization != null) {
             connectionURLs =
-                    getFallbackConnectionURL(
-                            connectionURLs,
-                            connectionURLPomData,
-                            config.getFallbackGitHubOrganization());
+                    getFallbackConnectionURL(connectionURLs, url, fallbackGitHubOrganization);
         }
 
-        IOException lastException = null;
+        PluginSourcesUnavailableException lastException = null;
```

Comment:
Following the general design of throwing up a checked exception for an "expected" failure (i.e., failure to clone or run Maven) but propagating `IOException` as `UncheckedIOException` for an "unexpected" failure (e.g., out of disk space).